src/cosl/coordinated_workers/nginx.py

```diff
@@ -29,6 +31,425 @@
+        except ModuleNotFoundError:
+            raise RuntimeError(
+                "Unmet dependency: the coordinated-workers NginxConfig object depends on the 'crossplane' package. \
+            Please install it with: `pip install crossplane` and add it to your charm's dependencies."
+            )
+        self._server_name = server_name
+        self._dns_IP_address = self._get_dns_ip_address()
+        self._ipv6_enabled = is_ipv6_enabled()
+        self._upstream_configs = upstream_configs
+        self._server_ports_to_locations = server_ports_to_locations
+        self._enable_health_check = enable_health_check
+        self._enable_status_page = enable_status_page
+
+    def get_config(self, addresses_by_role: Dict[str, Set[str]], tls: bool) -> str:
+        """Render the Nginx configuration as a string."""
+        full_config = self._prepare_config(addresses_by_role, tls)
+        return self._builder(full_config)  # type: ignore
+
+    def _prepare_config(
+        self, addresses_by_role: Dict[str, Set[str]], tls: bool
+    ) -> List[Dict[str, Any]]:
+        upstreams = self._upstreams(addresses_by_role)
+        # extract the upstream name
+        backends = [upstream["args"][0] for upstream in upstreams]
+        # build the complete configuration
+        full_config = [
+            {"directive": "worker_processes", "args": ["5"]},
+            {"directive": "error_log", "args": ["/dev/stderr", "error"]},
+            {"directive": "pid", "args": ["/tmp/nginx.pid"]},
+            {"directive": "worker_rlimit_nofile", "args": ["8192"]},
+            {
+                "directive": "events",
+                "args": [],
+                "block": [{"directive": "worker_connections", "args": ["4096"]}],
+            },
+            {
+                "directive": "http",
+                "args": [],
+                "block": [
+                    # upstreams (load balancing)
+                    *upstreams,
+                    # temp paths
+                    {
+                        "directive": "client_body_temp_path",
+                        "args": ["/tmp/client_temp"],
+                    },
+                    {"directive": "proxy_temp_path", "args": ["/tmp/proxy_temp_path"]},
+                    {"directive": "fastcgi_temp_path", "args": ["/tmp/fastcgi_temp"]},
+                    {"directive": "uwsgi_temp_path", "args": ["/tmp/uwsgi_temp"]},
+                    {"directive": "scgi_temp_path", "args": ["/tmp/scgi_temp"]},
+                    # logging
+                    {"directive": "default_type", "args": ["application/octet-stream"]},
+                    {
+                        "directive": "log_format",
+                        "args": [
+                            "main",
+                            '$remote_addr - $remote_user [$time_local]  $status "$request" $body_bytes_sent "$http_referer" "$http_user_agent" "$http_x_forwarded_for"',
+                        ],
+                    },
```

Comment:
I would not hardcode these values here, I would them as module's constants:

- "/dev/stderr"
- "/tmp/nginx.pid"
- "8192"
- "4096"
...